Firestore/Source/Core/FSTFirestoreClient.mm

```diff
@@ -230,9 +238,10 @@ - (void)disableNetworkWithCompletion:(nullable FSTVoidErrorBlock)completion {
   [self.workerDispatchQueue dispatchAsync:^{
     [self.remoteStore disableNetwork];
     if (completion) {
-      [self.userDispatchQueue dispatchAsync:^{
+      auto block = ^{
```

Comment:
This is surprisingly constructed. Why not call `completion` directly within the lambda submitted to the executor?

Here and throughout.

Reply:
Sorry, I meant to write a question on this. I'm not sure how C++ lambda and ARC play together. I understand that an Objective-C block must extend the lifetime of the objects it references, but I'm not sure if the same is true for a lambda, which is why I explicitly create a block. If that's not actually an issue, I'll be happy to just use a lambda.

Reply:
You can think of ARC as a compiler extension that inserts calls to `-retain` and `-release` whenever a pointer to Objective-C object is copied/goes out of scope. This interaction is preserved correctly within C++ too, so for example the default copy constructor of a C++ class containing an Objective-C object will do the right thing.

For example:
```c++
#import <Foundation/Foundation.h>

#include <iostream>

class Person {
 public:
  Person(NSString* name) : name_(name) {
  }

  NSString* name_;
};

void Print(Person person) {
  std::cerr << "Hello " << person.name_.UTF8String << std::endl;
}

int main() {
  Person person{@"Bob"};
  Print(person);
}
```
If you ask the compiler for assembly like so:

```
clang++ -S -fobjc-arc -std=c++11 test.mm
```

You can see the generated copy constructor includes a call to `_objc_retain`:

```asm
	.globl	__ZN6PersonC2ERKS_      ## -- Begin function _ZN6PersonC2ERKS_
	.weak_def_can_be_hidden	__ZN6PersonC2ERKS_
	.p2align	4, 0x90
__ZN6PersonC2ERKS_:                     ## @_ZN6PersonC2ERKS_
	.cfi_startproc
## BB#0:
	pushq	%rbp
Lcfi24:
	.cfi_def_cfa_offset 16
Lcfi25:
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
Lcfi26:
	.cfi_def_cfa_register %rbp
	subq	$32, %rsp
	movq	%rdi, -8(%rbp)
	movq	%rsi, -16(%rbp)
	movq	-8(%rbp), %rsi
	movq	-16(%rbp), %rdi
	movq	(%rdi), %rdi
	movq	%rsi, -24(%rbp)         ## 8-byte Spill
	callq	_objc_retain
	movq	-24(%rbp), %rsi         ## 8-byte Reload
	movq	%rax, (%rsi)
	addq	$32, %rsp
	popq	%rbp
	retq
	.cfi_endproc
```